### PR TITLE
[SPARK-48144][SQL] Fix `canPlanAsBroadcastHashJoin` to respect shuffle join hints

### PR DESCRIPTION
**Pull request by fred-db**
### What changes were proposed in this pull request?

* Currently, `canPlanAsBroadcastHashJoin` incorrectly returns that a join can be planned as a BHJ, even though the join contains a SHJ.
* To fix this, add some logic that checks whether the join contains a SHJ hint before checking if the join can be broadcasted.
* Also made a small refactor to the `JoinSelectionHelperSuite` to make it a bit more readable.

### Why are the changes needed?

* `canPlanAsBroadcastHashJoin` should be in sync with the join selection in `SparkStrategies`. Currently, it is not in sync.

### Does this PR introduce _any_ user-facing change?

Yes, semi / anti joins that could not have been planned as broadcasts would now not be pushed through aggregates anymore. Generally, this would be a performance improvement.

### How was this patch tested?

* Added UTs to check that a join with a SHJ hint is not marked as being planned as a BHJ.
* Added tests to keep `canPlanAsBroadcastHashJoin` and the `JoinSelection` codepath in sync.

### Was this patch authored or co-authored using generative AI tooling?

* No

### CODE REVIEW

**Review by agubichev — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 1]
```diff
@@ -191,6 +203,13 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
 //    ).foreach { case (query, joinClass) => assertJoin(query, joinClass) }
 //  }
 
+  test("broadcastable join with shuffle join hint") {
+    spark.sharedState.cacheManager.clearCache()
+    sql("CACHE TABLE testData")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:208]: why do we need to cache?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 2]
```diff
@@ -400,11 +395,32 @@ trait JoinSelectionHelper {
...
+        log"support binary equality. Keys not supporting hash joins: " +
+        log"${
+          MDC(HASH_JOIN_KEYS, keysNotSupportingHashJoin.map(
+            e => e.toString + " due to DataType: " + e.dataType.typeName).mkString(", "))
+        }")
+    }
+    result
+  }
+
+  def canPlanAsBroadcastHashJoin(join: Join, conf: SQLConf): Boolean = join match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:414]: to confirm: do we call this function in `JoinSelection`?

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
If you don't mind, can we proceed this parameter change separately before this PR, @fred-db ? That will help us review (or revert only the bug fix part later when we found some bugs in the future).

```
  def getBroadcastBuildSide(
-      left: LogicalPlan,
-      right: LogicalPlan,
-      joinType: JoinType,
-      hint: JoinHint,
+      join: Join,
```

**Comment by dongjoon-hyun** [comment]
Gentle ping, @fred-db .

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala (hunk 1 above)
Comment (fred-db) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:208]: That's how we made sure to make the table broadcastable, it's also used in some other tests.

**Comment by fred-db** [comment]
@dongjoon-hyun created the refactor PR here https://github.com/apache/spark/pull/46525. Let me know what you think! :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala (hunk 2 above)
Comment (fred-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:414]: No, the `JoinSelection` is unaffected. It's only called in `PushDownLeftSemiAntiJoin`

**Comment by dongjoon-hyun** [comment]
Thank you so much.

**Comment by dongjoon-hyun** [comment]
I merged the following. Could you rebase once more, @fred-db ?
- https://github.com/apache/spark/pull/46525.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs.) Thank you, @fred-db and all.

**Comment by fred-db** [comment]
Thanks @dongjoon-hyun ! :) Rebased the code.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.
Thank you, @fred-db , @agubichev,  @cloud-fan .

**Pull request closed without merging**